src/svc_ioq.c

```diff
@@ -179,66 +214,50 @@ ioq_flushv(SVCXPRT *xprt, struct x_vc_data *xd, struct xdr_ioq *xioq)
 	}
 }
 
-static void
-svc_ioq_callback(struct work_pool_entry *wpe)
+static inline void
+svc_ioq_loop(SVCXPRT *xprt, struct xdr_ioq *xioq, struct poolq_head *ifph)
 {
-	struct x_vc_data *xd = (struct x_vc_data *)wpe;
-	SVCXPRT *xprt = (SVCXPRT *)wpe->arg;
 	struct poolq_entry *have;
-	struct xdr_ioq *xioq;
 
-	/* qmutex more fine grained than xp_lock */
 	for (;;) {
-		mutex_lock(&xd->shared.ioq.qmutex);
-		have = TAILQ_FIRST(&xd->shared.ioq.qh);
-		if (unlikely(!have)) {
-			xd->shared.active = false;
-			mutex_unlock(&xd->shared.ioq.qmutex);
-			SVC_RELEASE(xprt, SVC_RELEASE_FLAG_NONE);
-			return;
-		}
-
-		TAILQ_REMOVE(&xd->shared.ioq.qh, have, q);
-		(xd->shared.ioq.qcount)--;
 		/* do i/o unlocked */
-		mutex_unlock(&xd->shared.ioq.qmutex);
-		xioq = _IOQ(have);
+		mutex_unlock(&ifph->qmutex);
 
 		if (svc_work_pool.params.thrd_max
 		 && !(xprt->xp_flags & SVC_XPRT_FLAG_DESTROYED)) {
 			/* all systems are go! */
-			ioq_flushv(xprt, xd, xioq);
+			svc_ioq_flushv(xprt, xioq);
 		}
+		SVC_RELEASE(xprt, SVC_RELEASE_FLAG_NONE);
 		XDR_DESTROY(xioq->xdrs);
-	}
 
-	return;
+		mutex_lock(&ifph->qmutex);
+		if (--(ifph->qcount) == 0)
+			break;
+
+		have = TAILQ_FIRST(&ifph->qh);
+		TAILQ_REMOVE(&ifph->qh, have, q);
+		xioq = _IOQ(have);
+		xprt = (SVCXPRT *)xioq->xdrs[0].x_lib[1];
+	}
 }
 
 void
-svc_ioq_append(SVCXPRT *xprt, struct x_vc_data *xd, XDR *xdrs)
+svc_ioq_append(SVCXPRT *xprt, struct xdr_ioq *xioq)
 {
-	if (unlikely(!svc_work_pool.params.thrd_max
-		  || (xprt->xp_flags & SVC_XPRT_FLAG_DESTROYED))) {
-		/* discard */
-		XDR_DESTROY(xdrs);
-		return;
-	}
+	struct poolq_head *ifph = &ioq_ifqh[xprt->xp_ifindex & IOQ_IF_MASK];
+
+	SVC_REF(xprt, SVC_REF_FLAG_NONE);
+	mutex_lock(&ifph->qmutex);
 
-	/* qmutex more fine grained than xp_lock */
-	mutex_lock(&xd->shared.ioq.qmutex);
-	(xd->shared.ioq.qcount)++;
-	TAILQ_INSERT_TAIL(&xd->shared.ioq.qh, &(XIOQ(xdrs)->ioq_s), q);
-
-	if (!xd->shared.active) {
-		xd->shared.active = true;
-		xd->wpe.fun = svc_ioq_callback;
-		xd->wpe.arg = xprt;
-		mutex_unlock(&xd->shared.ioq.qmutex);
-		SVC_REF(xprt, SVC_REF_FLAG_NONE);
-		work_pool_submit(&svc_work_pool, &xd->wpe);
+	if ((ifph->qcount)++ > 0) {
+		/* queue additional output requests without task switch */
+		xioq->xdrs[0].x_lib[1] = (void *)xprt;
+		TAILQ_INSERT_TAIL(&ifph->qh, &(xioq->ioq_s), q);
 	} else {
-		/* queuing multiple output requests for worker efficiency */
-		mutex_unlock(&xd->shared.ioq.qmutex);
+		/* handle additional output requests without task switch */
```

Comment:
This is a confusing comment to me, especially in conjunction with the one right above.  Maybe something like "handle the first output request inline"?